src/integritysetup/integritysetup-generator.c

```diff
@@ -0,0 +1,226 @@
+                arg_integritytab,
+                dd, dd);
+
+        fprintf(f,
+                "\n"
+                "[Service]\n"
+                "Type=oneshot\n"
+                "RemainAfterExit=yes\n"
+                "TimeoutSec=0\n"
+                "ExecStart=%s open '%s' '%s' %s\n"
+                "ExecStop=%s close '%s'\n",
+                arg_integritysetupbin, device, name_escaped, (options_convert) ? options_convert: "",
+                arg_integritysetupbin, name_escaped);
+
+        r = fflush_and_check(f);
+        if (r < 0)
+                return log_error_errno(r, "Failed to write unit file %s: %m", n);
+
+        r = generator_add_symlink(arg_dest, "integritysetup.target", "requires", n);
+        if (r < 0)
+                return r;
+
+        dmname = strjoina("dev-mapper-", e, ".device");
+        return generator_add_symlink(arg_dest, dmname, "requires", n);
+}
+
+static int add_integritytab_devices(void) {
+        _cleanup_fclose_ FILE *f = NULL;
+        unsigned integritytab_line = 0;
+        int r;
+
+        r = fopen_unlocked(arg_integritytab, "re", &f);
+        if (r < 0) {
+                if (errno != ENOENT)
+                        log_error_errno(errno, "Failed to open %s: %m", arg_integritytab);
+                return 0;
+        }
+
+        for (;;) {
+                _cleanup_free_ char *line = NULL, *name = NULL, *device_id = NULL, *device_path = NULL, *options = NULL;
+                char *l;
+
+                r = read_line(f, LONG_LINE_MAX, &line);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to read %s: %m", arg_integritytab);
+                if (r == 0)
+                        break;
+
+                integritytab_line++;
+
+                l = strstrip(line);
+                if (!l)
+                        continue;
+
+                if (IN_SET(l[0], 0, '#'))
+                        continue;
+
+                /* Options are optional */
+                r = sscanf(l, "%ms %ms %ms", &name, &device_id, &options);
+                if (!IN_SET(r, 2, 3)) {
+                        log_error("Failed to parse %s:%u, ignoring.", l, integritytab_line);
+                        continue;
+                }
+
+                device_path = fstab_node_to_udev_node(device_id);
+                if (!device_path) {
+                        log_error("Failed to find device %s:%u, ignoring.", device_id, integritytab_line);
+                        continue;
+                }
+
+                r = create_disk(name, device_path, options);
+                if (r < 0)
+                        return r;
+        }
+
+        return 0;
+}
+
+static int run(const char *dest, const char *dest_early, const char *dest_late) {
+        assert_se(arg_dest = dest);
+
+        arg_integritytab = getenv("SYSTEMD_INTEGRITYTAB") ?: "/etc/integritytab";
+        arg_integritysetupbin = getenv("SYSTEMD_INTEGRITYSETUPBIN") ?: "/usr/sbin/integritysetup";
```

Comment:
hmmm, I think for similar cases like this we called the variables `SYSTEMD_xyz_PATH` (and not `SYSTEMD_xyzBIN`). I think we should follow this scheme here too.

These

Reply:
So I appreciate the fact that you are just calling the upstream integritysetup binary here. While I sympathize with the simplicity I think we shouldn't do it that way. i.e. we should be very careful with defining the integritytab file format, but if we just mangle all options and directly reformat them into integritysetup cmdline switches without understanding what they do we are doing a bad job add stabilizing that file format.

So yeah, I think we should have systemd-integritysetup the same way as we have systemd-cryptsetup or systemd-veritysetup — it should do the basic setup we need, and support the precise list of options we intend to support, but not blanket pass them on.

I mean, sooner or later we have to understand things better anyway, since we ave to alter the deps we generate if people split out integrity data rom the main integrity volume.

Hence, in the interest of stabilizing the integritytab interface: let's do what we do for cryptsetup/veritysetup here, and have our own minimal wrapper around the lib, that supports only what we really want to support.

This should become pretty relevant as soon as we want to add some of the options that crypttab/veritytab so support, such as noauto/nofail/_netdev/x-initrd.attach/x-systemd.device-timeout= and so on.

But even withouth that: for keyed integrity partitions we need to supply the key somehow, and there we should really provide the same simple mechanisms as with cryptsetup, i.e. that you specify a file if you want, but that file can also be an AF_UNIX socket we read the key from. Also, if nothing is specified it searches for the key in /etc/cryptsetup-keys.d/. We should support the same here too (and it's very easy, given we already have the infra).

Now, I am not expecting that you add feature parity with the other two cases here (i.e. with crypttab/verittyab) in your patch. Not at all, we can add features as we progress. But I really want to make sure the format we definite with integrityttab supports only the features we really know we can support safely, and does not support more than we really know about. But by passing the arguments mapped 1:1 to cmdline switches this all is out of our control.

hence: please add a minimal systemd-integritysetup tool, that works ike systemd-veritysetup (you can probably copy 70% of the code) and just does the minimal set of features we want initially. Once we have that, we can add more features as we go.

Reply:
OK, next PR update will include the requested changes, thanks.